pkg/kubelet/kubelet_pods.go

```diff
@@ -1737,6 +1786,14 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(cgroupPods map[types.UID]cm.CgroupN
 			continue
 		}
 
+		// if the pod is within termination grace period, we shouldn't cleanup the underlying cgroup
+		kl.podKillingLock.Lock()
```

Comment:
`podKillingLock` and `mirrorPodTerminationMap` being accessible anywhere in the kubelet is messy.  We need an abstraction on top of it.

Reply:
something along the lines of
```
type PodCleanupManager struct {
  podManager
  podKillingLock
  mirrorPodTerminationMap
}

// GracePeriod struct can be made private to this

func RecordGracePeriod
func DeleteOrphanedMirrorPods
func IsPodTerminationPending
```

Reply:
@tedyu and I discussed something similar this morning.

I think an abstraction that lets us get visibility into the the pods queued up in the pod killer channel is the core need.  So rather that kubelet managing a podKillerCh, it could manage a `PodKiller` and support queue operations, and visibility into contents of the queue.  I see little reason to track a grace period time with pods in that interface for any reason other than logging in debug logs.  We just need to know if a pod or its mirror is in the queue for killing.

Reply:
hold on this for now though. i haven't fully thought it out.

Reply:
sounds like you and @derekwaynecarr where already going down this path :+1:

Reply:
@sjenning @derekwaynecarr 
I have refactored the code by creating PodKiller interface and PodKillerWithChannel struct.

Please take a look.

Thanks

Reply:
I have consolidated the locks in PodKillerWithChannel#PerformPodKillingWork() :
```
  	lock := sync.Mutex{}
```
the above lock from existing code and podKillingLock are combined into one lock.

Reply:
Potentially we can experiment with pod killer where the receipt of pod killing requests is via, e.g. slice (instead of channel).
But that requires more validation work and can be done in a subsequent PR.

Reply:
@sjenning @derekwaynecarr 
Please take another look.

Thanks